fourier/fourier.go

```diff
@@ -0,0 +1,254 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package fourier
+
+// FFT implements Fast Fourier Transform and its inverse for real sequences.
+type FFT struct {
+	work []float64
+	ifac [15]int
+
+	// real temporarily store complex data as
+	// pairs of real values to allow passing to
+	// the backing code. The length of real
+	// must always be half the length of work.
+	real []float64
+}
+
+// NewFFT returns an FFT initialized for work on sequences of length n.
+func NewFFT(n int) *FFT {
+	var t FFT
+	t.Reset(n)
+	return &t
+}
+
+// Len returns the length of the acceptable input.
+func (t *FFT) Len() int { return len(t.work) / 2 }
+
+// Reset reinitializes the FFT for work on sequences of length n.
+func (t *FFT) Reset(n int) {
+	if 2*n <= cap(t.work) {
+		t.work = t.work[:2*n]
+		t.real = t.real[:n]
+	} else {
+		t.work = make([]float64, 2*n)
+		t.real = make([]float64, n)
+	}
+	rffti(n, t.work, t.ifac[:])
+}
+
+// FFT computes the Fourier coefficients of the input sequence, seq,
+// placing the result in dst and returning it. This transform is
+// unnormalized since a call to FFT followed by a call of IFFT will
```

Comment:
"This transform is unnormalized, _so_ a call"?

Reply:
I'm reflecting the original FFTPACK documentation. I think it comes down to a question of causality; the behaviour of the implementation is that is causes a multiplication, so it is therefore unnormalised. The the alternative wording is that has unnormalised behaviour and so it is therefor causing a multiplication. I think the former is marginally tighter.

Reply:
maybe "in that"? Doesn't assign any causality, just a fact.

Reply:
I'm a fan of the ';' for this; that is a what a semicolon means.